File: agoritmos, todos/grafo dinamico y ordenamiento/grafoDinamico.py
import tkinter as tk
from tkinter import ttk, messagebox, scrolledtext
import heapq
import random
from collections import defaultdict
import math

class Nodo:
    def __init__(self, nombre, x, y):
        self.nombre = nombre
        self.x = x
        self.y = y

class Arista:
    def __init__(self, desde, hasta, peso):
        self.desde = desde
        self.hasta = hasta
        self.peso = peso

class GrafoDinamico:
    def __init__(self, root):
        self.root = root
        self.root.title("Grafo Transporte - Entrada Manual")
        self.root.geometry("800x600")
        
        self.nodos = {}
        self.aristas = []
        self.ruta_corta = []
        
        self.setup_ui()
        
    def setup_ui(self):
        # Frame superior para entrada de datos
        top_frame = ttk.Frame(self.root)
        top_frame.pack(fill='x', padx=10, pady=10)
        
        # Label y área de texto para aristas
        ttk.Label(top_frame, text="Ingrese aristas (formato: Nodo1 Nodo2 Peso):").pack(anchor='w')
        
        entrada_frame = ttk.Frame(top_frame)
        entrada_frame.pack(fill='x', pady=5)
        
        self.entrada_text = scrolledtext.ScrolledText(entrada_frame, height=4, width=60)
        self.entrada_text.pack(side='left', fill='both', expand=True)
        self.entrada_text.insert('1.0', "A B 4\nA C 2\nB D 10\nC D 3\nC E 8\nD E 6")
        
        self.crear_grafo_button = ttk.Button(entrada_frame, text="Crear Grafo", command=self.cargar_grafo_desde_texto)
        self.crear_grafo_button.pack(side='right', padx=(10, 0))
        
        # Frame para controles de búsqueda
        control_frame = ttk.Frame(self.root)
        control_frame.pack(fill='x', padx=10, pady=5)
        
        ttk.Label(control_frame, text="Inicio:").pack(side='left')
        self.inicio_field = ttk.Entry(control_frame, width=10)
        self.inicio_field.pack(side='left', padx=5)
        
        ttk.Label(control_frame, text="Fin:").pack(side='left')
        self.fin_field = ttk.Entry(control_frame, width=10)
        self.fin_field.pack(side='left', padx=5)
        
        self.buscar_ruta_button = ttk.Button(control_frame, text="Buscar Ruta", command=self.buscar_ruta)
        self.buscar_ruta_button.pack(side='left', padx=10)
        
        # Área de resultado
        resultado_frame = ttk.Frame(self.root)
        resultado_frame.pack(fill='x', padx=10, pady=5)
        
        self.resultado_text = scrolledtext.ScrolledText(resultado_frame, height=3, width=60)
        self.resultado_text.pack(fill='x')
        
        # Canvas para dibujar el grafo
        self.canvas = tk.Canvas(self.root, bg='white', height=300)
        self.canvas.pack(fill='both', expand=True, padx=10, pady=10)
        
    def cargar_grafo_desde_texto(self):
        self.nodos.clear()
        self.aristas.clear()
        self.ruta_corta.clear()
        
        lineas = self.entrada_text.get('1.0', tk.END).strip().split('\n')
        
        for linea in lineas:
            parts = linea.strip().split()
            if len(parts) == 3:
                try:
                    n1, n2, peso = parts[0].upper(), parts[1].upper(), int(parts[2])
                    
                    # Crear nodos si no existen
                    if n1 not in self.nodos:
                        self.nodos[n1] = Nodo(n1, random.randint(100, 500), random.randint(100, 250))
                    if n2 not in self.nodos:
                        self.nodos[n2] = Nodo(n2, random.randint(100, 500), random.randint(100, 250))
                    
                    # Crear aristas bidireccionales
                    self.aristas.append(Arista(self.nodos[n1], self.nodos[n2], peso))
                    self.aristas.append(Arista(self.nodos[n2], self.nodos[n1], peso))
                    
                except ValueError:
                    continue
        
        self.dibujar_grafo()
    
    def dijkstra(self, inicio):
        distancias = {nodo: float('inf') for nodo in self.nodos.values()}
        padres = {}
        distancias[inicio] = 0
        padres[inicio] = None
        
        cola = [(0, inicio)]
        
        while cola:
            dist_actual, nodo_actual = heapq.heappop(cola)
            
            if dist_actual > distancias[nodo_actual]:
                continue
            
            for arista in self.aristas:
                if arista.desde == nodo_actual:
                    nueva_dist = dist_actual + arista.peso
                    if nueva_dist < distancias[arista.hasta]:
                        distancias[arista.hasta] = nueva_dist
                        padres[arista.hasta] = nodo_actual
                        heapq.heappush(cola, (nueva_dist, arista.hasta))
        
        return distancias, padres
    
    def reconstruir_ruta(self, destino, padres):
        ruta = []
        nodo_actual = destino
        
        while nodo_actual is not None:
            ruta.append(nodo_actual)
            nodo_actual = padres.get(nodo_actual)
        
        ruta.reverse()
        return ruta
    
    def buscar_ruta(self):
        inicio_texto = self.inicio_field.get().upper()
        fin_texto = self.fin_field.get().upper()
        
        if inicio_texto not in self.nodos or fin_texto not in self.nodos:
            messagebox.showerror("Error", "Nodos inválidos")
            return
        
        nodo_inicio = self.nodos[inicio_texto]
        nodo_fin = self.nodos[fin_texto]
        
        distancias, padres = self.dijkstra(nodo_inicio)
        
        if distancias[nodo_fin] == float('inf'):
            self.resultado_text.delete('1.0', tk.END)
            self.resultado_text.insert('1.0', "No hay camino")
            self.ruta_corta = []
        else:
            distancia_total = distancias[nodo_fin]
            self.ruta_corta = self.reconstruir_ruta(nodo_fin, padres)
            
            resultado = "Ruta más corta: "
            for nodo in self.ruta_corta:
                resultado += nodo.nombre + " "
            resultado += f"\nDistancia: {distancia_total}"
            
            self.resultado_text.delete('1.0', tk.END)
            self.resultado_text.insert('1.0', resultado)
        
        self.dibujar_grafo()
    
    def dibujar_grafo(self):
        self.canvas.delete("all")
        
        # Dibujar aristas
        aristas_dibujadas = set()
        for arista in self.aristas:
            # Evitar dibujar la misma arista dos veces
            arista_key = tuple(sorted([arista.desde.nombre, arista.hasta.nombre]))
            if arista_key not in aristas_dibujadas:
                aristas_dibujadas.add(arista_key)
                
                self.canvas.create_line(arista.desde.x, arista.desde.y, 
                                      arista.hasta.x, arista.hasta.y,
                                      fill='gray', width=1)
                
                # Dibujar peso en el punto medio
                xm = (arista.desde.x + arista.hasta.x) // 2
                ym = (arista.desde.y + arista.hasta.y) // 2
                self.canvas.create_text(xm, ym, text=str(arista.peso), fill='black')
        
        # Dibujar ruta más corta
        if len(self.ruta_corta) > 1:
            for i in range(len(self.ruta_corta) - 1):
                n1 = self.ruta_corta[i]
                n2 = self.ruta_corta[i + 1]
                self.canvas.create_line(n1.x, n1.y, n2.x, n2.y,
                                      fill='red', width=3)
        
        # Dibujar nodos
        for nodo in self.nodos.values():
            # Círculo del nodo
            self.canvas.create_oval(nodo.x - 15, nodo.y - 15, 
                                  nodo.x + 15, nodo.y + 15,
                                  fill='#66ccff', outline='black', width=2)
            
            # Nombre del nodo
            self.canvas.create_text(nodo.x, nodo.y, text=nodo.nombre, 
                                  fill='black', font=('Arial', 10, 'bold'))

def main():
    root = tk.Tk()
    app = GrafoDinamico(root)
    root.mainloop()

if __name__ == "__main__":
    main()

File: agoritmos, todos/grafo dinamico y ordenamiento/grafo_dinamico.cpp
#include <iostream>
#include <vector>
#include <map>
#include <queue>
#include <string>
#include <sstream>
#include <algorithm>
#include <climits>
#include <random>
#include <FL/Fl.H>
#include <FL/Fl_Window.H>
#include <FL/Fl_Button.H>
#include <FL/Fl_Input.H>
#include <FL/Fl_Text_Editor.H>
#include <FL/Fl_Text_Buffer.H>
#include <FL/Fl_Box.H>
#include <FL/fl_draw.H>
#include <FL/Fl_Widget.H>

using namespace std;

struct Nodo {
    string nombre;
    int x, y;
    
    Nodo(const string& n, int x_pos, int y_pos) : nombre(n), x(x_pos), y(y_pos) {}
};

struct Arista {
    Nodo* desde;
    Nodo* hasta;
    int peso;
    
    Arista(Nodo* d, Nodo* h, int p) : desde(d), hasta(h), peso(p) {}
};

struct NodoDistancia {
    Nodo* nodo;
    int distancia;
    
    NodoDistancia(Nodo* n, int d) : nodo(n), distancia(d) {}
    
    bool operator>(const NodoDistancia& other) const {
        return distancia > other.distancia;
    }
};

class GrafoDinamico : public Fl_Widget {
private:
    map<string, unique_ptr<Nodo>> nodos;
    vector<unique_ptr<Arista>> aristas;
    vector<Nodo*> rutaCorta;
    
    Fl_Text_Buffer* entradaBuffer;
    Fl_Text_Editor* entradaTextArea;
    Fl_Input* inicioField;
    Fl_Input* finField;
    Fl_Button* crearGrafoButton;
    Fl_Button* buscarRutaButton;
    Fl_Text_Buffer* resultadoBuffer;
    Fl_Text_Editor* resultadoArea;
    
    random_device rd;
    mt19937 gen;
    
public:
    GrafoDinamico(int x, int y, int w, int h) : Fl_Widget(x, y, w, h), gen(rd()) {
        setupUI();
    }
    
    void setupUI() {
        // Configurar área de entrada
        entradaBuffer = new Fl_Text_Buffer();
        entradaBuffer->text("A B 4\nA C 2\nB D 10\nC D 3\nC E 8\nD E 6");
        
        // Configurar área de resultado
        resultadoBuffer = new Fl_Text_Buffer();
    }
    
    void cargarGrafoDesdeTexto() {
        nodos.clear();
        aristas.clear();
        rutaCorta.clear();
        
        string texto = entradaBuffer->text();
        istringstream iss(texto);
        string linea;
        
        uniform_int_distribution<> randX(100, 600);
        uniform_int_distribution<> randY(100, 400);
        
        while (getline(iss, linea)) {
            if (linea.empty()) continue;
            
            istringstream lineStream(linea);
            string n1, n2;
            int peso;
            
            if (lineStream >> n1 >> n2 >> peso) {
                transform(n1.begin(), n1.end(), n1.begin(), ::toupper);
                transform(n2.begin(), n2.end(), n2.begin(), ::toupper);
                
                if (nodos.find(n1) == nodos.end()) {
                    nodos[n1] = make_unique<Nodo>(n1, randX(gen), randY(gen));
                }
                if (nodos.find(n2) == nodos.end()) {
                    nodos[n2] = make_unique<Nodo>(n2, randX(gen), randY(gen));
                }
                
                aristas.push_back(make_unique<Arista>(nodos[n1].get(), nodos[n2].get(), peso));
                aristas.push_back(make_unique<Arista>(nodos[n2].get(), nodos[n1].get(), peso));
            }
        }
    }
    
    map<Nodo*, int> dijkstra(Nodo* inicio, map<Nodo*, Nodo*>& padres) {
        map<Nodo*, int> distancias;
        priority_queue<NodoDistancia, vector<NodoDistancia>, greater<NodoDistancia>> cola;
        
        distancias[inicio] = 0;
        padres[inicio] = nullptr;
        cola.push(NodoDistancia(inicio, 0));
        
        while (!cola.empty()) {
            NodoDistancia actual = cola.top();
            cola.pop();
            
            Nodo* vActual = actual.nodo;
            int distActual = actual.distancia;
            
            if (distancias.count(vActual) && distancias[vActual] < distActual)
                continue;
            
            for (const auto& a : aristas) {
                if (a->desde == vActual) {
                    int nuevaDist = distActual + a->peso;
                    if (!distancias.count(a->hasta) || nuevaDist < distancias[a->hasta]) {
                        distancias[a->hasta] = nuevaDist;
                        padres[a->hasta] = vActual;
                        cola.push(NodoDistancia(a->hasta, nuevaDist));
                    }
                }
            }
        }
        
        return distancias;
    }
    
    vector<Nodo*> reconstruirRuta(Nodo* destino, const map<Nodo*, Nodo*>& padres) {
        vector<Nodo*> ruta;
        for (Nodo* v = destino; v != nullptr; v = padres.count(v) ? padres.at(v) : nullptr) {
            ruta.push_back(v);
        }
        reverse(ruta.begin(), ruta.end());
        return ruta;
    }
    
    void buscarRuta() {
        string inicioTexto = inicioField->value();
        string finTexto = finField->value();
        
        transform(inicioTexto.begin(), inicioTexto.end(), inicioTexto.begin(), ::toupper);
        transform(finTexto.begin(), finTexto.end(), finTexto.begin(), ::toupper);
        
        if (nodos.find(inicioTexto) == nodos.end() || nodos.find(finTexto) == nodos.end()) {
            resultadoBuffer->text("Nodos inválidos");
            return;
        }
        
        Nodo* startNode = nodos[inicioTexto].get();
        Nodo* endNode = nodos[finTexto].get();
        
        map<Nodo*, Nodo*> padres;
        auto distancias = dijkstra(startNode, padres);
        
        if (distancias.find(endNode) == distancias.end()) {
            resultadoBuffer->text("No hay camino");
            rutaCorta.clear();
        } else {
            int distanciaTotal = distancias[endNode];
            rutaCorta = reconstruirRuta(endNode, padres);
            
            string resultado = "Ruta más corta: ";
            for (const auto& n : rutaCorta) {
                resultado += n->nombre + " ";
            }
            resultado += "\nDistancia: " + to_string(distanciaTotal);
            resultadoBuffer->text(resultado.c_str());
        }
        
        redraw();
    }
    
    void draw() override {
        fl_rectf(x(), y(), w(), h(), FL_WHITE);
        
        // Dibujar aristas
        fl_color(FL_GRAY);
        fl_line_style(FL_SOLID, 1);
        
        for (const auto& a : aristas) {
            fl_line(a->desde->x, a->desde->y, a->hasta->x, a->hasta->y);
            
            int xm = (a->desde->x + a->hasta->x) / 2;
            int ym = (a->desde->y + a->hasta->y) / 2;
            
            fl_color(FL_BLACK);
            fl_draw(to_string(a->peso).c_str(), xm, ym);
            fl_color(FL_GRAY);
        }
        
        // Dibujar ruta más corta
        if (rutaCorta.size() > 1) {
            fl_color(FL_RED);
            fl_line_style(FL_SOLID, 3);
            
            for (size_t i = 0; i < rutaCorta.size() - 1; i++) {
                Nodo* n1 = rutaCorta[i];
                Nodo* n2 = rutaCorta[i + 1];
                fl_line(n1->x, n1->y, n2->x, n2->y);
            }
        }
        
        // Dibujar nodos
        for (const auto& pair : nodos) {
            Nodo* n = pair.second.get();
            
            fl_color(fl_rgb_color(102, 204, 255));
            fl_pie(n->x - 15, n->y - 15, 30, 30, 0, 360);
            
            fl_color(FL_BLACK);
            fl_circle(n->x - 15, n->y - 15, 30);
            fl_draw(n->nombre.c_str(), n->x - 5, n->y + 5);
        }
        
        fl_line_style(FL_SOLID, 1);
    }
    
    void setEntradaBuffer(Fl_Text_Buffer* buffer) { entradaBuffer = buffer; }
    void setResultadoBuffer(Fl_Text_Buffer* buffer) { resultadoBuffer = buffer; }
    void setInicioField(Fl_Input* field) { inicioField = field; }
    void setFinField(Fl_Input* field) { finField = field; }
};

// Funciones callback
void crear_grafo_callback(Fl_Widget* widget, void* data) {
    GrafoDinamico* grafo = static_cast<GrafoDinamico*>(data);
    grafo->cargarGrafoDesdeTexto();
    grafo->redraw();
}

void buscar_ruta_callback(Fl_Widget* widget, void* data) {
    GrafoDinamico* grafo = static_cast<GrafoDinamico*>(data);
    grafo->buscarRuta();
}

int main() {
    Fl_Window window(800, 600, "Grafo Transporte - Entrada Manual");
    
    // Área de entrada
    Fl_Text_Buffer* entradaBuffer = new Fl_Text_Buffer();
    entradaBuffer->text("A B 4\nA C 2\nB D 10\nC D 3\nC E 8\nD E 6");
    
    Fl_Text_Editor* entradaArea = new Fl_Text_Editor(10, 30, 500, 100);
    entradaArea->buffer(entradaBuffer);
    
    Fl_Button* crearButton = new Fl_Button(520, 30, 100, 30, "Crear Grafo");
    
    // Campos de entrada
    Fl_Input* inicioField = new Fl_Input(70, 140, 50, 25, "Inicio:");
    Fl_Input* finField = new Fl_Input(180, 140, 50, 25, "Fin:");
    Fl_Button* buscarButton = new Fl_Button(240, 140, 100, 25, "Buscar Ruta");
    
    // Área de resultado
    Fl_Text_Buffer* resultadoBuffer = new Fl_Text_Buffer();
    Fl_Text_Editor* resultadoArea = new Fl_Text_Editor(10, 170, 500, 60);
    resultadoArea->buffer(resultadoBuffer);
    
    // Área de dibujo
    GrafoDinamico* grafoPanel = new GrafoDinamico(10, 240, 780, 350);
    grafoPanel->setEntradaBuffer(entradaBuffer);
    grafoPanel->setResultadoBuffer(resultadoBuffer);
    grafoPanel->setInicioField(inicioField);
    grafoPanel->setFinField(finField);
    
    // Configurar callbacks
    crearButton->callback(crear_grafo_callback, grafoPanel);
    buscarButton->callback(buscar_ruta_callback, grafoPanel);
    
    window.end();
    window.show();
    
    return Fl::run();
}